### DISCUSSION TITLE
MouseMoved, MouseIn and MouseOut do not fire. Any ideas why?

### DISCUSSION THREAD

**Discussion by buffos** [fyne-io/fyne#3964] [Q&A] [answered]
I am using fyne/v2  **v2.3.5** . I am trying to learn it.

The simple widget extends the BaseWidget.

I am implementing **MouseMoved MouseIn **and** MouseOut** (simply some print lines to verify they are called). Nothing is ever printed

On the other hand, MouseDown and MouseUp event are being triggered.

Do you have any ideas why that's happening? I tried different layouts, only by itself, those MouseMoved events will not get fired.

**I am using:**
Windows 11
tdm-gcc 10.3.0
go 1.20

**Comment by Jacalz** [comment]
I think this is almost impossible to answer without seeing the code. Would you mind sharing a simple example?

**Reply by buffos** [reply]
Here is a minimal code that reproduces the problem I face. A minimal custom widget.
Mousedown and mouse up are triggered. Not MouseMove, In, Out.

```golang
package main

import (
	"fmt"
	"image/color"

	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/canvas"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/driver/desktop"
	"fyne.io/fyne/v2/widget"
)

type TestWidget struct {
	widget.BaseWidget
	Text string
}

func NewTestWidget(text string) *TestWidget {
	w := &TestWidget{}
	w.Text = text
	w.ExtendBaseWidget(w)
	return w
}

func (w *TestWidget) CreateRenderer() fyne.WidgetRenderer {
	text := canvas.NewText(w.Text, color.Black)
	objects := []fyne.CanvasObject{text}
	return &TestWidgetRenderer{text, objects, w}
}

func (w *TestWidget) MouseDown(ev *desktop.MouseEvent) {
	fmt.Println("Mouse Down")
}

func (w *TestWidget) MouseUp(ev *desktop.MouseEvent) {
	fmt.Println("Mouse Up")
}

func (w *TestWidget) MouseIn(ev *desktop.MouseEvent) {
	fmt.Println("Mouse In")
}

func (w *TestWidget) MouseOut(ev *desktop.MouseEvent) {
	fmt.Println("Mouse In")
}

func (w *TestWidget) MouseMoved(ev *desktop.MouseEvent) {
	fmt.Println("Mouse Moved")
}

type TestWidgetRenderer struct {
	text    *canvas.Text
	objects []fyne.CanvasObject
	parent  *TestWidget
}

func (r *TestWidgetRenderer) Destroy() {
}

func (r *TestWidgetRenderer) Layout(size fyne.Size) {
	r.text.Resize(size)
}

func (r *TestWidgetRenderer) MinSize() fyne.Size {
	return r.text.MinSize()
}

func (r *TestWidgetRenderer) Objects() []fyne.CanvasObject {
	return r.objects
}

func (r *TestWidgetRenderer) Refresh() {
	r.text.Text = r.parent.Text
	r.text.Refresh()
}

func main() {
	myApp := app.New()
	label := NewTestWidget("Hello World!")
	container.NewVBox(label)
	myWindow := myApp.NewWindow("Label Widget")
	myWindow.SetContent(container.NewVBox(label))
	myWindow.ShowAndRun()
}
```

**Reply by buffos** [reply]
Never mind I found the problem. 

I was passing an argument to  MouseOut, thus failing to implement the Hoverable interface and making it not work.

Thanks anyone for their time.

**Reply by Jacalz** [reply, marked as answer]
I'm glad to hear that you got it working. I would suggest adding compile-time checks for interface implementations in that case. Add this to get compiler errors if the widget doesn't implement the specified interface:
```go
var _ desktop.Hoverable = (*TestWidget)(nil)
```
